### PR TITLE
Merge ITextAndVersion and ITextVersionable

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
No need to for separate interfaces here.  ITextAndVersion  should just expose the 'try' version of "give me the version stamp", just like it exposes the async version of it.

### CODE REVIEW

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 1]
```diff
@@ -143,7 +139,7 @@ public async ValueTask<VersionStamp> GetVersionAsync(LoadTextOptions options, Ca
             return recoverableText.Version;
         }
 
-        private sealed class RecoverableText : WeaklyCachedRecoverableValueSource<SourceText>, ITextVersionable
+        private sealed class RecoverableText : WeaklyCachedRecoverableValueSource<SourceText>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:142]: didn't need this interface here.   RecoverableText is never exposed outside of RecoverableTextAndVersion.  So everything in RecoverableTextAndVersion just casts directly to RecoverableText and can grab anything it needs off of that.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 2]
```diff
@@ -73,14 +71,12 @@ public bool TryGetValue(LoadTextOptions options, [MaybeNullWhen(false)] out Text
 
         public bool TryGetTextVersion(LoadTextOptions options, out VersionStamp version)
         {
-            if (_initialSourceOrRecoverableText is ITextVersionable textVersionable)
-            {
-                return textVersionable.TryGetTextVersion(options, out version);
-            }
+            if (TryGetInitialSourceOrRecoverableText(out var source, out var recoverableText))
+                return source.TryGetTextVersion(options, out version);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:75]: defer to common helper which either gives the original 'source' or the 'recoverable text'.  if the former wins, we just defer to that to try to get the version.  This also moves to match the pattern of the method above.

File: src/Workspaces/Core/Portable/Workspace/Solution/ITextAndVersionSource.cs [hunk 3]
```diff
@@ -20,6 +20,8 @@ internal interface ITextAndVersionSource
     TextAndVersion GetValue(LoadTextOptions options, CancellationToken cancellationToken);
     Task<TextAndVersion> GetValueAsync(LoadTextOptions options, CancellationToken cancellationToken);
 
+    bool TryGetTextVersion(LoadTextOptions options, out VersionStamp version);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/ITextAndVersionSource.cs:23]: 'Try' version of the GetVersionAsync method below.

File: src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs [hunk 4]
```diff
@@ -57,11 +57,9 @@ private bool TryGetInitialSourceOrRecoverableText([NotNullWhen(true)] out ITextA
         public bool TryGetValue(LoadTextOptions options, [MaybeNullWhen(false)] out TextAndVersion value)
         {
             if (TryGetInitialSourceOrRecoverableText(out var source, out var recoverableText))
-            {
                 return source.TryGetValue(options, out value);
-            }
 
-            if (recoverableText.TryGetValue(out var text) && recoverableText.LoadTextOptions == options)
+            if (recoverableText.LoadTextOptions == options && recoverableText.TryGetValue(out var text))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:62]: reversed order.  matches code in the method below, and allows us to fast path out.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs [hunk 5]
```diff
@@ -113,24 +113,7 @@ public bool TryGetText([NotNullWhen(returnValue: true)] out SourceText? text)
...
-                version = textAndVersion.Version;
-                return true;
-            }
-            else
-            {
-                version = default;
-                return false;
-            }
-        }
+            => TextAndVersionSource.TryGetVersion(LoadTextOptions, out version);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentState.cs:116]: much simpler now, and matches method below.

File: src/Workspaces/Core/Portable/Workspace/Solution/ITextAndVersionSource.cs [hunk 6]
```diff
@@ -20,6 +20,8 @@ internal interface ITextAndVersionSource
     TextAndVersion GetValue(LoadTextOptions options, CancellationToken cancellationToken);
     Task<TextAndVersion> GetValueAsync(LoadTextOptions options, CancellationToken cancellationToken);
 
+    bool TryGetVersion(LoadTextOptions options, out VersionStamp version);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/ITextAndVersionSource.cs:23]: 'Try' version of the GetVersionAsync method below.

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski ptal.

**Review by jasonmalinowski — APPROVED** [review]
This seems suspiciously simple. But nice since it's hard to tell if the old code would have worked or accidentally had some type _fail_ to implement the interface you deleted and it wouldn't work.

**Comment by CyrusNajmabadi** [comment]
Thanks! :)

**Pull request merged**